File: app/src/main/java/oleg_pronin/tetris/models/Block.java
package oleg_pronin.tetris.models;

import android.graphics.Color;
import android.graphics.Point;

import androidx.annotation.NonNull;

import java.util.Random;

import oleg_pronin.tetris.constants.FieldConstants;

public class Block{
    private final int shapeIndex;
    private int frameNumber;

    private final BlockColor color;
    private Point position;

    public Block(int shapeIndex, BlockColor color) {
        this.frameNumber = 0;
        this.shapeIndex = shapeIndex;
        this.color = color;
        this.position = new Point(FieldConstants.COLUMN_COUNT.getValue() / 2, 0);
    }

    public static Block createBlock() {
        Random random = new Random();
        int shapeIndex = random.nextInt(Shape.values().length);
        BlockColor blockColor = BlockColor.values()[random.nextInt(BlockColor.values().length)];

        Block block = new Block(shapeIndex, blockColor);
        block.position.x = block.position.x - Shape.values()[shapeIndex].getStartPosition();

        return block;
    }

    public static int getColor(byte value) {
        for (BlockColor color: BlockColor.values()) {
            if (value == color.byteValue) {
                return  color.rgbValue;
            }
        }

        return -1;
    }

    public final void setState(int frame, Point position) {
        this.frameNumber = frame;
        this.position = position;
    }

    @NonNull
    public final byte[][] getShape(int frameNumber) {
        return Shape.values()[shapeIndex].getFrame(frameNumber).as2dByteArray();
    }

    public Point getPosition() {
        return position;
    }

    public final int getFrameCount() {
        return Shape.values()[shapeIndex].getFrameCount();
    }

    public int getFrameNumber() {
        return frameNumber;
    }

    public int getColor() {
        return color.rgbValue;
    }

    public byte getStaticValue() {
        return color.byteValue;
    }

    public enum BlockColor {
        PINK(Color.rgb(255, 105, 180), (byte) 2),
        GREEN(Color.rgb(0, 128, 0), (byte) 3),
        ORANGE(Color.rgb(255, 140, 0), (byte) 4),
        YELLOW(Color.rgb(255, 255, 0), (byte) 5),
        CYAN(Color.rgb(0, 255, 255), (byte) 6);

        BlockColor(int rgbValue, byte byteValue) {
            this.rgbValue = rgbValue;
            this.byteValue = byteValue;
        }

        private final int rgbValue;
        private final byte byteValue;
    }
}

File: app/src/main/java/oleg_pronin/tetris/models/Block.kt
package oleg_pronin.tetris.models

import android.graphics.Color
import android.graphics.Point
import oleg_pronin.tetris.constants.FieldConstants
import java.util.*

class Block(private val shapeIndex: Int, private val color: BlockColor) {
    private var frameNumber = 0
    private var position: Point

    init {
        position = Point(FieldConstants.COLUMN_COUNT.value / 2, 0)
    }

    fun setState(frame: Int, position: Point) {
        frameNumber = frame
        this.position = position
    }

    fun getShape(frameNumber: Int): Array<ByteArray> {
        return Shape.values()[shapeIndex].getFrame(frameNumber).as2dByteArray()
    }

    fun getColor(): Int {
        return color.rgbValue
    }

    val staticValue: Byte
        get() = color.byteValue.toByte()

    val frameCount: Int
        get() = Shape.values()[shapeIndex].frameCount

    enum class BlockColor(val rgbValue: Int, byteValue: Byte) {
        PINK(Color.rgb(255, 105, 180), 2.toByte()),
        GREEN(Color.rgb(0, 128, 0), 3.toByte()),
        ORANGE(Color.rgb(255, 140, 0), 4.toByte()),
        YELLOW(Color.rgb(255, 255, 0), 5.toByte()),
        CYAN(Color.rgb(0, 255, 255), 6.toByte());

        val byteValue: Int = byteValue.toInt()
    }

    companion object {
        fun createBlock(): Block {
            val random: Random = Random()
            val shareIndex: Int = random.nextInt(Shape.values().size)
            val blockColor: BlockColor = BlockColor.values()[random.nextInt(BlockColor.values().size)]
            val block: Block = Block(shareIndex, blockColor)

            block.position.x = block.position.x - Shape.values()[shareIndex].startPosition

            return block
        }

        fun getColor(value: Byte): Int {
            for (color in BlockColor.values()) {
                if (value.toInt() == color.byteValue) {
                    return color.rgbValue
                }
            }

            return -1
        }
    }
}